p2p/src/peer_manager/peerdb/address_data/mod.rs

```diff
@@ -162,20 +163,27 @@ impl AddressData {
         )
     }
 
-    fn next_connect_time(
-        now: Duration,
-        fail_count: u32,
-        reserved: bool,
-        rng: &mut impl Rng,
-    ) -> Duration {
-        now + Self::next_connect_delay(fail_count, reserved)
-            .mul_f64(utils::exp_rand::exponential_rand(rng))
+    fn next_connect_time(now: Time, fail_count: u32, reserved: bool, rng: &mut impl Rng) -> Time {
+        let random_offset = Self::next_connect_delay(fail_count, reserved)
+            .mul_f64(utils::exp_rand::exponential_rand(rng));
+        loop {
+            let res = now + random_offset;
+            match res {
+                Some(r) => return r,
+                None => {
+                    // The random offset is too large, try again
+                    logging::log::debug!(
+                        "next_connect_time: Random_offset too large (now: {now:?}, offset: {random_offset:?}), trying again"
+                    );
+                }
+            }
+        }
     }
```

Comment:
This loop looks a bit unnecessary from the one hand and scary from the other.
1) The addition won't fail if `next_connect_delay` and `exponential_rand` behave as expected, because the returned numbers are small enough.
2) If, on the other hand, we assume that those functions, in particular `next_connect_delay`, may return something unexpected, then the loop may become infinite.

Reply:
The reason why I used a loop here is that the value of totally probabilistic. I can't prove this won't produce an infinitely large value. Statistically, it's possible to prove the standard deviation is bounded. But for singular values, the standard deviation is virtually infinite, so it can be anything.

Reply:
I mean, as I can see, the result of `exponential_rand` is either relatively small or is infinity (because it's minus logarithm of a number from [0,1), so its largest finite value is `-f64::MIN_POSITIVE.ln()` which is `708.3964185322641`). Also, if it's infinite, `mul_f64` will panic, so the loop won't have a chance to fix this problem.
And `next_connect_delay` returns at most `MAX_DELAY_REACHABLE`, which is 1 hour. So, under normal circumstances the loop won't do anything (unless I'm missing something of course).
And if we expect some "abnormal" circumstances, it still won't solve all the problems, e.g. if `exponential_rand` could return arbitrarily large finite values, `next_connect_time` could produce a value which is billions of years into the future.

I'd suggest that we `clamp` the result of `exponential_rand` here instead, so that it's guaranteed to be reasonable.
(It also makes sense to fix `exponential_rand` so that it always return finite values. But `next_connect_time` should probably still not rely on the values being small).

Reply:
Nothing prevents the random number from being large enough to cause an overflow in addition, but small enough to not trigger the mul_f64 overflow. That's totally possible. On the other hand, statistically speaking, the loop will never iterate more than once, if ever...

Clamping sounds like a reasonable solution. The only issue with it is that it breaks encapsulation. I'm happy to add a TODO to fix this later. Also, if you know how to fix this in a few minutes + add a test to ensure the correct behavior, please feel free to do it, and push a commit here.